galley/pkg/server/components/validation.go

```diff
@@ -15,36 +15,167 @@
+// This is for lint fix
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+func monitorReadiness(stop <-chan struct{}, port uint, readiness *probe.Probe) {
+	const httpsHandlerReadinessFreq = time.Second
+
+	ready := false
+	client := &http.Client{
+		Timeout: time.Second,
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		},
+	}
+
+	for {
+		if err := webhookHTTPSHandlerReady(client, port); err != nil {
+			readiness.SetAvailable(errors.New("not ready"))
+			scope.Infof("https handler for validation webhook is not ready: %v\n", err)
+			ready = false
+		} else {
+			readiness.SetAvailable(nil)
+			if !ready {
+				scope.Info("https handler for validation webhook is ready\n")
+				ready = true
+			}
+		}
+
+		select {
+		case <-stop:
+			return
+		case <-time.After(httpsHandlerReadinessFreq):
+			// check again
+		}
+	}
+}
+
+func webhookHTTPSHandlerReady(client httpClient, port uint) error {
+	readinessURL := &url.URL{
+		Scheme: "https",
+		Host:   fmt.Sprintf("localhost:%v", port),
+		Path:   server.HTTPSHandlerReadyPath,
+	}
+
+	req := &http.Request{
+		Method: http.MethodGet,
+		URL:    readinessURL,
+	}
+
+	response, err := client.Do(req)
+	if err != nil {
+		return fmt.Errorf("HTTP request to %v failed: %v", readinessURL, err)
+	}
+	defer response.Body.Close()
+	if response.StatusCode != http.StatusOK {
+		return fmt.Errorf("GET %v returned non-200 status=%v",
+			readinessURL, response.StatusCode)
+	}
+	return nil
+}
+
+func NewValidationServer(
+	options server.Options,
+	liveness probe.Controller,
+	readiness probe.Controller,
+) process.Component {
+	return process.ComponentFromFns(
+		// start
+		func() error {
+			stop := make(chan struct{})
+
+			log.Infof("Galley validation server started with \n%s", &options)
+
+			options.MixerValidator = validate.NewDefaultValidator(false)
+			options.PilotDescriptor = schemas.Istio
+			s, err := server.New(options)
+			if err != nil {
+				return fmt.Errorf("cannot create validation webhook service: %v", err)
+			}
+
+			if options.Mux == nil {
+				if liveness != nil {
+					validationLivenessProbe := probe.NewProbe()
+					validationLivenessProbe.SetAvailable(nil)
+					validationLivenessProbe.RegisterProbe(liveness, "validationLiveness")
+
+					go func() {
+						<-stop
+						validationLivenessProbe.SetAvailable(errors.New("stopped"))
+					}()
+				}
+
+				if readiness != nil {
+					validationReadinessProbe := probe.NewProbe()
+					validationReadinessProbe.SetAvailable(errors.New("init"))
+					validationReadinessProbe.RegisterProbe(readiness, "validationReadiness")
+
+					go func() {
+						monitorReadiness(stop, options.Port, validationReadinessProbe)
+
+						validationReadinessProbe.SetAvailable(errors.New("stopped"))
+					}()
+				}
+			}
 
+			go s.Run(stop)
+			go cmd.WaitSignal(stop)
+			return nil
+		},
+		// stop
+		func() {
+			// server doesn't have a stop function.
```

Comment:
What about close the stop channel? right now it seems like that is never done?

Reply:
Good point. This was carried over from the existing code which seemed to rely on the process being aborted. This probably worked in the normal case but would cause problems if we need to re-start a component for testing purposes.

Reply:
Will you address in this PR?

Reply:
Yes, I'll fix in this PR.